sdk/src/azure/core/az_mqtt5_rpc_client_codec.c

```diff
@@ -1,83 +1,420 @@
 #include <azure/core/az_result.h>
 #include <azure/core/internal/az_log_internal.h>
+#include <azure/core/internal/az_precondition_internal.h>
+#include <azure/core/internal/az_result_internal.h>
+#include <azure/iot/az_iot_common.h>
 #include <stdlib.h>
 
 #include <azure/core/_az_cfg.h>
 
+static const az_span az_mqtt5_rpc_client_codec_default_publish_topic_format
+    = AZ_SPAN_LITERAL_FROM_STR("services/{serviceId}/{executorId}/command/{name}/request");
+static const az_span az_mqtt5_rpc_client_codec_default_subscribe_topic_format
+    = AZ_SPAN_LITERAL_FROM_STR(
+        "clients/{invokerClientId}/services/{serviceId}/{executorId}/command/{name}/response");
+static const az_span az_mqtt5_rpc_client_id_key = AZ_SPAN_LITERAL_FROM_STR("{invokerClientId}");
+static const az_span az_mqtt5_rpc_service_id_key = AZ_SPAN_LITERAL_FROM_STR("{serviceId}");
+static const az_span az_mqtt5_rpc_executor_id_key = AZ_SPAN_LITERAL_FROM_STR("{executorId}");
+static const az_span az_mqtt5_rpc_command_id_key = AZ_SPAN_LITERAL_FROM_STR("{name}");
+static const az_span az_mqtt5_rpc_any_executor_id = AZ_SPAN_LITERAL_FROM_STR("_any_");
+
 AZ_NODISCARD az_mqtt5_rpc_client_codec_options az_mqtt5_rpc_client_codec_options_default()
 {
-  return (az_mqtt5_rpc_client_codec_options){ .subscribe_timeout_in_seconds
-                                              = AZ_MQTT5_RPC_DEFAULT_TIMEOUT_SECONDS,
-                                              .publish_timeout_in_seconds
-                                              = AZ_MQTT5_RPC_DEFAULT_TIMEOUT_SECONDS,
-                                              .subscription_topic_format = AZ_SPAN_EMPTY,
-                                              .request_topic_format = AZ_SPAN_EMPTY };
+  return (az_mqtt5_rpc_client_codec_options){
+    .subscribe_timeout_in_seconds = AZ_MQTT5_RPC_DEFAULT_TIMEOUT_SECONDS,
+    .publish_timeout_in_seconds = AZ_MQTT5_RPC_DEFAULT_TIMEOUT_SECONDS,
+    .subscription_topic_format = az_mqtt5_rpc_client_codec_default_subscribe_topic_format,
+    .request_topic_format = az_mqtt5_rpc_client_codec_default_publish_topic_format
+  };
 }
 
-AZ_NODISCARD az_result az_mqtt5_rpc_client_codec_get_subscription_topic(
+AZ_NODISCARD az_result az_mqtt5_rpc_client_codec_get_publish_topic(
     az_mqtt5_rpc_client_codec* client,
-    az_span out_subscription_topic,
-    int32_t* out_topic_length)
+    az_span executor_id,
+    az_span command_name,
+    char* mqtt_topic,
+    size_t mqtt_topic_size,
+    size_t* out_mqtt_topic_length)
 {
-  return az_rpc_get_topic_from_format(
-      client->_internal.options.subscription_topic_format,
-      client->_internal.model_id,
-      AZ_SPAN_FROM_STR("+"),
-      client->_internal.client_id,
-      _az_span_is_valid(client->_internal.command_name, 1, 0) ? client->_internal.command_name
-                                                              : AZ_SPAN_FROM_STR("+"),
-      out_subscription_topic,
-      out_topic_length);
+  _az_PRECONDITION_NOT_NULL(client);
```

Comment:
Could these get topic functions use a single underlying function with different parameters? They seem very similar

Reply:
Will implement in a follow-up PR!